Change JSON structure in `testimony print`

mirzal opened the pull request:
`testimony --json print` was direct dump of internal testimony data structures. It was a little awkward to work with from external point of view.

- new structure is simplified - it's shallow list of objects with tokens and some additional fields
- include test name and class for each test case

GitHub diff view is misleading a bit, I moved ` if SETTINGS['json']:` part above main printing part.

mirzal commented:
New JSON structure:
```json
[
    {
        "requirement": "Location",
        "caseautomation": "Automated",
        "caselevel": "Acceptance",
        "casecomponent": "OrganizationsLocations",
        "testtype": "Functional",
        "caseimportance": "Critical",
        "upstream": "No",
        "id": "90bb90a3-120f-4ea6-89a9-62757be42486",
        "expectedresults": "Location created successfully and has expected and\ncorrect name",
        "test": "Create new locations using different inputs as a name",
        "test-path": "tests/foreman/api/test_location.py",
        "test-class": "LocationTestCase",
        "test-name": "test_positive_create_with_name",
        "testimony-invalid-tokens": {},
        "testimony-rst-parse-messages": []
    }
]
```

With #145, it may also include `decorators`, which is a list of decorators merged from class level and function level (`decorators` key may not be present at all).

`test-class` might contain null (`None` in Python).

@jhutar Is this something that you will find easy to work with in reportportal_cli, or would you prefer some changes to structure?
You can compute Report Portal id using `test-path`, `test-class` and `test-name`. This is simple list of objects, so you can use list comprehensions to filter out irrelevant items before processing them.

sthirugn commented:
It looks good to me. Nice work!

mirzal commented:
> LGTM. We currently don't have the JSON output being tested when running `make test` maybe it worth adding that.

How blocking that is for you?
At the very least, I would prefer we finalize JSON output first, and add a test or two later. Doesn't make sense for me to add tests when JSON structure is still discussed.

> Another thing to consider is that with this flat structure, it is hard to figure what is token and what is meta information provided by testimony. The ones prefixed with `testimony-` make that easy but `test-class` is something that testimony collects but is not a token.

Would it work for you if we add `testimony-` prefix to everything provided by testimony, so `testimony-test-path`, `testimony-test-class` and `testimony-test-name`?

Or would you rather go back to nested structure, something like:
```
{"test-path": "foo/bar.py",
 "test-class": "TestClass",
 "invalid-tokens": {},
 "tokens": {
    "caseautomation": "Automated"
 }
}
```
or even
```
{"caseautomation": "Automated",
 "requirement": "Location",
 "testimony-metadata": {
    "test-path": "foo/bar.py"
  }
}
```

> Talking about testing testimony, we could try to write some tests using pytest for it in the future. This would ensure it is working fine. Click offers good tooling for testing CLI apps written using it.

I agree on principle. Testimony accumulated sizable technical debt. We should introduce proper unit testing and refactor current solution heavily. Unfortunately, there are always more pressing matters on our priorities list...

elyezer commented:
@mirzal JSON output testing is not blocking anything. I suggested that since we are now caring about its structure so having tests would help us ensure the schema does not break.

About the JSON structure I don't have a personal opinion but I like the idea of nesting testimony metadata into its own dict. That said I would suggest the following structure:

```json
{
  "field1": "value1",
  "field2": "value2",
  "fieldN": "valueN",
  "_metadata": {
    "test-path": "test/path",
    "other-meta: "",
}
```

With the above structure Testimony would not override any of the user defined fields. `_metadata` may be something that someone would use as a field, so maybe we could use `_testimony` instead? What do you think?

As I said I don't have a personal preference but I feel like most of the time someone will be trying to access the field information rather than testimony metadata. Because of that that I think the metadata should be nested instead of the fields and that a name that wouldn't likely conflict with a user defined field name should be used for it.

Feel free to pick one that fits better for the use case you have in mind. Thank you very much for improving the JSON output it was indeed messy.

elyezer commented:
https://github.com/SatelliteQE/testimony/pull/145 is now merged. Where are you thinking about adding them? Into the metadata?

mirzal commented:
@elyezer I rebased on top of current master and changed structure as you suggested. Right now testimony outputs this:

```json
[
    {
        "requirement": "Location",
        "caseautomation": "Automated",
        "caselevel": "Integration",
        "casecomponent": "OrganizationsLocations",
        "testtype": "Functional",
        "caseimportance": "Critical",
        "upstream": "No",
        "id": "2786146f-f466-4ed8-918a-5f46806558e2",
        "expectedresults": "Location updated successfully and has correct capsule\nassigned",
        "bz": "1398695",
        "test": "Update location with new capsule",
        "decorators": [
            "run_in_one_thread",
            "tier2"
        ],
        "_testimony": {
            "file-path": "tests/foreman/api/test_location.py",
            "test-class": "LocationTestCase",
            "test-name": "test_positive_create_update_and_remove_capsule",
            "invalid-tokens": {},
            "rst-parse-messages": []
        }
    },
]
```

As for decorators - Satellite uses them for tiers, which are metadata, but they may also contain things that can't be considered metadata. For that reason, I feel they should belong to `_testimony`. However, previous PR already put them in `tokens`. I don't like idea of moving them into `_testimony` just for printing. Should I move them out of tokens into class property? In this PR, or separate one? Or just leave it as is?

Decorators were put into `tokens`, so they can be included in `testimony print` output. With that PR, it doesn't matter anymore for `--json`, but is still something to consider in human-readable output.

elyezer commented:
@mirzal I think it is fine to leave `tokens` where they are now. Let's leave it there, since it is a new feature I think it is fine. If you feel that it would make more sense moving to metadata, feel free to do so but this is up to you.

Thank you
